mobile_verifier/src/subsciber_location.rs

```diff
@@ -0,0 +1,179 @@
+    ) -> Self {
+        Self {
+            pool,
+            carrier_keys_refresh_interval: settings.carrier_keys_refresh_interval(),
+            carrier_keys,
+            _config_client: config_client,
+            reports_receiver,
+        }
+    }
+    pub async fn run(mut self, shutdown: &triggered::Listener) -> anyhow::Result<()> {
+        let mut refresh_carrier_keys_timer = time::interval(self.carrier_keys_refresh_interval);
+        refresh_carrier_keys_timer.set_missed_tick_behavior(MissedTickBehavior::Skip);
+        loop {
+            if shutdown.is_triggered() {
+                break;
+            }
+            tokio::select! {
+                _ = shutdown.clone() => break,
+                _ = refresh_carrier_keys_timer.tick() =>
+                match self.refresh_carrier_keys().await {
+                    Ok(()) => (),
+                    Err(err) => {
+                        tracing::error!("error whilst refreshing carrier keys: {err:?}");
+                    }
+                },
+
+                msg = self.reports_receiver.recv() => if let Some(stream) =  msg {
+                    self.process_file(stream).await?;
+                }
+            }
+        }
+        tracing::info!("stopping subscriber location reports handler");
+        Ok(())
+    }
+
+    async fn process_file(
+        &self,
+        file_info_stream: FileInfoStream<SubscriberLocationIngestReport>,
+    ) -> anyhow::Result<()> {
+        let mut transaction = self.pool.begin().await?;
+        file_info_stream
+            .into_stream(&mut transaction)
+            .await?
+            .map(anyhow::Ok)
+            .try_fold(transaction, |mut transaction, report| async move {
+                if self.verify_known_carrier_key(&report.report.pubkey) {
+                    self.save(
+                        report.report.subscriber_id.clone(),
+                        report.received_timestamp,
+                        &mut transaction,
+                    )
+                    .await?;
+                    metrics::increment_counter!(
+                        "oracles_mobile_verifier_valid_subscriber_location_report"
+                    );
+                } else {
+                    metrics::increment_counter!(
+                        "oracles_mobile_verifier_invalid_subscriber_location_report"
+                    );
+                }
+                Ok(transaction)
+            })
+            .await?
+            .commit()
+            .await?;
+        Ok(())
+    }
+
+    pub async fn save(
+        &self,
+        subscriber_id: Vec<u8>,
+        timestamp: DateTime<Utc>,
+        db: &mut Transaction<'_, Postgres>,
+    ) -> Result<(), sqlx::Error> {
+        sqlx::query(
+            r#"
+                INSERT INTO subscriber_loc (subscriber_id, date_bucket, hour_bucket, reward_timestamp)
+                VALUES ($1, DATE_TRUNC('hour', $2), DATE_PART('hour', $2), $2)
```

Comment:
i think this could be simplified to just 2-3 columns: the subscriber ID, and the timestamp, or a separate "processing/ingest" timestamp from the reward timestamp if there's a meaningful distinction to be had there.
my understanding of the `date_trunc` and the `date_part` functions is that they are used to retrieve sub-pieces of the datetime timestamp field, but they still all operate on a timestamptz that's stored in the db. so you would just insert the timestamptz in these save functions but then you could retrieve different components (hour component) in the select query when you're retrieving records to do reward eligibility calculation